### PR TITLE
Make API feedback private

### LINKED ISSUE #2849: Make API feedback private
# Description

Based on customer feedback, some API owners may want either all or part of feedback be private.

# Goal

There is a setting
* [ ] to set all received feedback to be private
* [ ] to set an individual feedback item to be private
* [ ] When the feedback is set to private, it is still visible for
 * [ ] admin users
 * [ ] the API owner
 * [ ] the user who created the feedback (applies only to the feedback given by that particular user)

### PR DESCRIPTION
**Pull request by Vellames**
#Closes #2849 

WIP

Database:
- [x] Create 'isPublic' flag in feedback's schema
- [x] Create Migrations to ensure that all feedbacks have the 'isPublic' flag

Interface: 
- [x] Create button to mark all feedbacks as private
- [x] Create button to mark all feedbacks as public
- [x] Create button to mark a specific feedback as public
- [x] Create button to mark a specific feedback as private
- [x] Create icon to private feedbacks

Hide the privates feedbacks, except to:

- [x] Admin
- [x] API Owner
- [x] The User Who Created the Feedback

Permissions:
- [x] API Owner
- [x] Admin

### CODE REVIEW

**Comment by Nazarah** [comment]
@Vellames whenever there is some UI ready, please attach screenshot for it so that we can give our feedback about the look and feel and the functionality.
Also make sure that to API owner, look and feel of private feedback should appear different than that of a public feedback.

**Comment by Vellames** [comment]
feedback @Nazarah Ok, this is the feedback item with the new funcionality

**Comment by Nazarah** [comment]
@Vellames please consider using either of the following options
1. using [fa-eye-slash](http://fontawesome.io/icon/eye-slash/) icon on a feedback that has been made private. The icon can appear after the field **Topic**
or
2. add a semi-transparent layer over a private comment to make it distinguishable from public ones.

**Comment by Nazarah** [comment]
@Vellames can you please show me the how a normal user sees his/her feedback that has been made private by the API owner?

**Comment by Vellames** [comment]
feedback

@Nazarah This is the feedback pannel with your solicitations

**Comment by Vellames** [comment]
> @Vellames can you please show me the how a normal user sees his/her feedback that has been made  private by the API owner?

There are not differences in visualization between a public/private feedback for the owner of it

**Comment by Vellames** [comment]
@Nazarah, @marla-singer  I think that it's all done, can you review please?

**Comment by Nazarah** [comment]
@Vellames Feedbacks
1. Rename the global buttons as "Make all feedback private" and "Make all feedback public".
2. Keep some space between those two buttons. Right now, the border style and placement of the buttons make them appear as one
3. Place the "Make feedback public" and "Make feedback private" between Edit and Delete buttons of a feedback. In this case, make the border style such that buttons can be distinguished from one another.

**Comment by Vellames** [comment]
@Nazarah 

I added a modal when the user click in mark all feedbacks as public/private:


modal

Normal user view of feedbacks:


userfeedback

Admin user view of feedbacks:


adminfeedback

**Comment by Nazarah** [comment]
@Vellames Looks good.
Just need to confirm this.
You showed the normal user view.
This is the user who made a feedback and api owner made that feedback private.
So normal user won't be seeing that his feedback has been made private. Am I right?

**Comment by Vellames** [comment]
> So normal user won't be seeing that his feedback has been made private. Am I right? 

Exactly

> Please made the modal text as "Are you sure to change all feedbacks to public?"

Ok

**Comment by Vellames** [comment]
@marla-singer Can you review the code please?

File: apinf_packages/core/migrations/server/9-add-ispublic-flag-in-feedback.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import { Migrations } from 'meteor/percolate:migrations';
+
+import Feedback from '/apinf_packages/feedback/collection';
+
+Migrations.add({
+  version: 9,
+  name: 'Ensure that all feedbacks have the isPublic flag',
+  up () {
+    Feedback.update({}, { $set: { isPublic: true } });
```
Comment (marla-singer) [inline apinf_packages/core/migrations/server/9-add-ispublic-flag-in-feedback.js]: Change query. Select all feedback docs where field `isPublic` doesn't exist.

File: apinf_packages/core/migrations/server/9-add-ispublic-flag-in-feedback.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import { Migrations } from 'meteor/percolate:migrations';
```
Comment (marla-singer) [inline apinf_packages/core/migrations/server/9-add-ispublic-flag-in-feedback.js]: Add license text

File: apinf_packages/feedback/client/change_visibility/change_visibility.html [hunk 3]
```diff
@@ -0,0 +1,47 @@
+<!-- Copyright 2017 Apinf Oy
+This file is covered by the EUPL license.
+You may obtain a copy of the licence at
+https://joinup.ec.europa.eu/community/eupl/og_page/european-union-public-licence-eupl-v11 -->
+
+<template name="changeFeedbackVisibility">
+    <div
```
Comment (marla-singer) [inline apinf_packages/feedback/client/change_visibility/change_visibility.html]: **Make** indent is 2 spaces, not 4

File: apinf_packages/feedback/client/change_visibility/change_visibility.html [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        class="modal fade"
+        tabindex="-1"
+        role="dialog"
+        aria-labelledby="changeFeedbackVisibilityItemModalLabel">
+        <div class="modal-dialog" role="document">
+        <div class="modal-content">
```
Comment (marla-singer) [inline apinf_packages/feedback/client/change_visibility/change_visibility.html]: Nested element has to have a new line

File: apinf_packages/feedback/client/change_visibility/change_visibility.html [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+            <div class="modal-header">
+            <button
```
Comment (marla-singer) [inline apinf_packages/feedback/client/change_visibility/change_visibility.html]: Nested elements have to have a new line

File: apinf_packages/feedback/client/change_visibility/change_visibility.html [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+                type="button"
+                class="close"
+                data-dismiss="modal"
+                aria-label="Close">
+                <span aria-hidden="true">&times;</span>
+                </button>
+            <h2 class="modal-title" id="changeFeedbackVisibilityItemModalTitle">
```
Comment (marla-singer) [inline apinf_packages/feedback/client/change_visibility/change_visibility.html]: 1. Why do need ID for `h2` tag? **Remove ID**

File: apinf_packages/feedback/client/change_visibility/change_visibility.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+https://joinup.ec.europa.eu/community/eupl/og_page/european-union-public-licence-eupl-v11 */
+
+// Meteor packages imports
+import { Meteor } from 'meteor/meteor';
+import { Template } from 'meteor/templating';
+
+// Meteor contributed packages imports
+import { Modal } from 'meteor/peppelg:bootstrap-3-modal';
+
+import { FlowRouter } from 'meteor/kadira:flow-router';
```
Comment (marla-singer) [inline apinf_packages/feedback/client/change_visibility/change_visibility.js]: 1. Delete an empty line. Package `kadira:flow-router` is related to Contributed packages imports block
2. Keep an alphabet order

File: apinf_packages/feedback/client/change_visibility/change_visibility.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+Template.changeFeedbackVisibility.events({
+  'click #confirm-change-visibility': function () {
```
Comment (marla-singer) [inline apinf_packages/feedback/client/change_visibility/change_visibility.js]: **Change** to an arrow style
`'click #selector': () => {}`

File: apinf_packages/feedback/client/change_visibility/change_visibility.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    // Get slug
+    const slug = FlowRouter.getParam('slug');
+
+    // Call meteor method
+    Meteor.call('changeAllFeedbacksVisibility', slug, this.toPublic);
```
Comment (marla-singer) [inline apinf_packages/feedback/client/change_visibility/change_visibility.js]: Describe in comment what does this method do?

File: apinf_packages/feedback/client/change_visibility/change_visibility.html [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+                {{_ "feedback_change_visibility_header"}}
+            </h2>
+            </div>
+            <div class="modal-body">
+                {{#if toPublic}}
```
Comment (marla-singer) [inline apinf_packages/feedback/client/change_visibility/change_visibility.html]: Spaces between the elements and `{{` `}}`
Must be `{{ #if toPublic }}`

File: apinf_packages/feedback/client/change_visibility/change_visibility.html [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+            <div class="modal-header">
+            <button
+                type="button"
+                class="close"
+                data-dismiss="modal"
+                aria-label="Close">
+                <span aria-hidden="true">&times;</span>
+                </button>
+            <h2 class="modal-title" id="changeFeedbackVisibilityItemModalTitle">
+                {{_ "feedback_change_visibility_header"}}
```
Comment (marla-singer) [inline apinf_packages/feedback/client/change_visibility/change_visibility.html]: Add a spacebar before `}}`

File: apinf_packages/feedback/client/change_visibility/change_visibility.html [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+            </h2>
+            </div>
+            <div class="modal-body">
+                {{#if toPublic}}
+                    {{_ "feedback_changeAllVisibility_toPublic"}}
```
Comment (marla-singer) [inline apinf_packages/feedback/client/change_visibility/change_visibility.html]: Add a spacebar before `}}`

File: apinf_packages/feedback/client/change_visibility/change_visibility.html [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+                {{else}}
+                    {{_ "feedback_changeAllVisibility_toPrivate"}}
```
Comment (marla-singer) [inline apinf_packages/feedback/client/change_visibility/change_visibility.html]: Add a spacebar before `}}`

File: apinf_packages/feedback/client/change_visibility/change_visibility.html [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+                <span aria-hidden="true">&times;</span>
+                </button>
+            <h2 class="modal-title" id="changeFeedbackVisibilityItemModalTitle">
+                {{_ "feedback_change_visibility_header"}}
+            </h2>
+            </div>
+            <div class="modal-body">
+                {{#if toPublic}}
+                    {{_ "feedback_changeAllVisibility_toPublic"}}
+                {{else}}
```
Comment (marla-singer) [inline apinf_packages/feedback/client/change_visibility/change_visibility.html]: Spaces
`{{ else }}`

File: apinf_packages/feedback/client/feedback.html [hunk 15]
```diff
@@ -7,14 +7,30 @@
...
+      <div id="feedback-options">
+        <button
+          type="button"
+          class="btn btn-info btn-xs pull-right feedback-form-button"
+          id="add-feedback">
+            <i class="fa fa-comments-o"></i>
+            &nbsp;
+            {{_ "feedback_feedbackForm_button" }}
+        </button>
+        {{#if api.currentUserCanManage}}
```
Comment (marla-singer) [inline apinf_packages/feedback/client/feedback.html]: Spaces
`{{ element }}`

File: apinf_packages/feedback/client/feedback.html [hunk 16]
```diff
@@ -7,14 +7,30 @@
...
+          <button 
+            type="button"
+            class="btn btn-info btn-xs pull-right feedback-form-button"
+            id="mark-all-feedbacks-as-private">
+              {{_ "feedback_mark_all_feedbacks_as_private"}}
```
Comment (marla-singer) [inline apinf_packages/feedback/client/feedback.html]: Add a spacebar before `}}`

File: apinf_packages/feedback/client/feedback.html [hunk 17]
```diff
@@ -7,14 +7,30 @@
...
+          </button>
+          <button 
+            type="button"
+            class="btn btn-info btn-xs pull-right feedback-form-button"
+            id="mark-all-feedbacks-as-public">
+            {{_ "feedback_mark_all_feedbacks_as_public"}}
```
Comment (marla-singer) [inline apinf_packages/feedback/client/feedback.html]: Add a spacebar before `}}`

File: apinf_packages/feedback/client/feedback.less [hunk 18]
```diff
@@ -0,0 +1,7 @@
+#feedback-visibility-options{
```
Comment (marla-singer) [inline apinf_packages/feedback/client/feedback.less]: Add a spacebar before `{`

File: apinf_packages/feedback/client/feedback.less [hunk 19]
```diff
@@ -0,0 +1,7 @@
...
+    margin-right: 1em;
+}
+
+#feedback-options button.feedback-form-button{
```
Comment (marla-singer) [inline apinf_packages/feedback/client/feedback.less]: Add a spacebar before {

File: apinf_packages/feedback/client/item/item.html [hunk 20]
```diff
@@ -4,59 +4,75 @@
...
-          <span id="down-vote-{{ item._id }}" class="fa fa-lg fa-arrow-down down-vote {{ userDownvote }}"></span>
-        </div>
-        <div class="col-md-9 form-inline">
-          <div class="form-group feedback-type">
-            <span>
-              <strong>{{_ "feedbackItem_type" }}:&nbsp;</strong>
-            </span>
-            <span class="badge">
-              {{ feedbackType }}
+  {{# if item.currentUserCanSee}}
```
Comment (marla-singer) [inline apinf_packages/feedback/client/item/item.html]: Spaces
`{{ element }}`

File: apinf_packages/feedback/client/item/item.html [hunk 21]
```diff
@@ -4,59 +4,75 @@
...
+            </div>
+            <div class="form-group feedback-topic">
+              <span>
+                <strong>{{_ "feedbackItem_topic" }}:&nbsp;</strong>
+              </span>
+              <span>
+                {{ item.topic }}
+              </span>
+              {{#unless item.isPublic}}
+                {{#if item.currentUserCanChangeVisibility}}
```
Comment (marla-singer) [inline apinf_packages/feedback/client/item/item.html]: Spaces
{{ element }}

**Comment by Vellames** [comment]
@marla-singer I did the changes, ready for review

**Pull request merged**